### PR TITLE
WIP: Basic data types for the Nova implementation

### PR DESCRIPTION
**Pull request by DanielRrr**
Here I introduce Lean counterparts of the data types used in the Rust implementation of the Nova SNARK. The description will be updated occasionally

### CODE REVIEW

File: Nova/SNARK/Circuit.lean [hunk 1]
```diff
@@ -0,0 +1,42 @@
+import Nova.SNARK.Commitments
+import Nova.SNARK.Errors
+import Nova.SNARK.R1CS
+import YatimaStdLib.Either
+
+-- A helper trait for a step of the incremental computation (i.e., circuit for F)
+class StepCircuit (F : Type _) where
```
Comment (mpenciak) [inline Nova/SNARK/Circuit.lean:7]: We may want to constrain `F` here to be the underlying field?

File: Nova/BellPerson/Solver.lean [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+  bv : Array Bit
+  total_density : USize
+
+def new_tracker : DensityTracker := DensityTracker.mk #[] 0
+
+def add_element (tr : DensityTracker) : DensityTracker :=
+  let new_bv :=  Array.push tr.bv .zero
+  DensityTracker.mk new_bv tr.total_density
+
+structure SatisfyingAssignment (G : Type _) where
```
Comment (mpenciak) [inline Nova/BellPerson/Solver.lean:13]: Do we want to constrain `G` here to be of the elliptic curve's group type?

**Review by mpenciak — CHANGES_REQUESTED** [review]
I think we should try to constrain the types of the `F`s and `G`s appearing as parameters in the definitions, see the more in-depth Zulip post I made.

File: Nova/BellPerson/Solver.lean (hunk 2 above)
Comment (DanielRrr) [reply Nova/BellPerson/Solver.lean:13]: Why don't we leave this type an arbitrary and declare that `G` is a group whenever we need that?

Comment (DanielRrr) [reply Nova/BellPerson/Solver.lean:13]: The same for the comment above

**Review by mpenciak — APPROVED** [review]

File: Nova/SNARK/R1CS.lean [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+  (u₁ : RelaxedR1CSInstance G) 
+  (u₂ : R1CSInstance G) 
+  (comm_T : Commitment G) (r : G) : Either Error (RelaxedR1CSInstance G) :=
+  let (x₁, u', comm_W₁, comm_E₁) := (u₁.X, u₁.u, u₁.comm_W, u₁.comm_E)
+  let (x₂, comm_W₂) := (u₂.X, u₂.comm_W)
+  let comm_W := (comm_W₁.comm : G) + ((comm_W₂.comm : G) * r)
+  let comm_E := comm_E₁.comm + comm_T.comm * r
+  let u := u' + r
+  let x := Array.map (fun (a,b) => a + b * r) (Array.zip x₁ x₂)
+  let result :=
```
Comment (DanielRrr) [inline Nova/SNARK/R1CS.lean:105]: @arthurpaulino, see this

**Pull request merged**